### PR TITLE
Add seed keyword argument to random_layout and spring_layout

### PR DESCRIPTION
**Pull request by harryscholes**
This PR adds functionality that allows reproducible node layouts to be generated by `random_layout` and `spring_layout` via the `seed` keyword argument.

---

Re-running code that employs stochastic layout algorithms to generate figures positions nodes differently each time:

```
import networkx as nx, matplotlib.pyplot as plt

G = nx.erdos_renyi_graph(10, 0.4, seed=1)

fig, axes = plt.subplots(3, 3)
axes = axes.flatten()

for ax in axes:
    pos = nx.spring_layout(G)
    nx.draw_networkx(G, pos, ax=ax, node_size=150, font_color='w')
    ax.axis('off')
```

random

This PR adds the `state` keyword argument to the `random_layout` and `spring_layout` functions to set the seed of `numpy.random.seed`. Setting `seed` allows reproducible node layouts to be generated:

```
fig, axes = plt.subplots(3, 3)
axes = axes.flatten()

for ax in axes:
    pos = nx.spring_layout(G, random_seed=1)
    nx.draw_networkx(G, pos, ax=ax, node_size=150, font_color='w')
    ax.axis('off')
```
seeded

This feature will be very useful when generating figures for presentations/papers.

### CODE REVIEW

**Comment by ghoresh11** [comment]
Great idea! This will be very useful

**Comment by dschult** [comment]
This looks good.  But could you take a look at #1764 and #1550 where it is argued that we shouldn't be doing this? What is the right approach to handling random number generators and numpy.random in particular?

**Comment by harryscholes** [comment]
@dschult thanks for these comments. As I have currently implemented this, the random state is set globally and leaks out of the scope of `spring_layout` and `random_layout`.

I have fixed this by providing a `random_state` argument that accepts an instance of a NumPy random number generator, an `int`, or `None`. If an `int` is provided, this will be used to instantiate `np.random.RandomState`. If `None` or `np.random`, this will be used to instantiate an unseeded `np.random.RandomState`.

```python
if random_state is None or random_state is np.random:
    rng = np.random.mtrand._rand
elif isinstance(random_state, np.random.RandomState):
    rng = random_state
else:
    rng = np.random.RandomState(random_state)
```
This is heavily inspired by how sklearn [does this](https://github.com/scikit-learn/scikit-learn/blob/a24c8b464d094d2c468a16ea9f8bf8d42d949f84/sklearn/utils/validation.py#L617).

Example :

```python
In [6]: nx.spring_layout(G, seed=1)
   ...: np.random.rand(5)
   ...:
Out[6]: array([ 0.84351284,  0.68102688,  0.11044101,  0.92682998,  0.81743538])

In [7]: nx.spring_layout(G, seed=1)
   ...: np.random.rand(5)
   ...:
Out[7]: array([ 0.72168598,  0.08398381,  0.82544323,  0.49460946,  0.62301562])
```

The same approach could be taken to fix #1764.

**Comment by dschult** [comment]
Very nice!   
This could be helpful elsewhere I'm sure... in addition to #1764 but certainly there.

It'd be nice, (for backward compatibility even for private functions) to leave the existing arguments in the same order and add the random state argument to the end. But if there are other reasons to add it earlier I'm open to that.  Also it should probably have a default -- perhaps of None -- even in private functions.

Does it make sense to pull together some helper code into the nx.utils to help with random states? I suppose it could even be a decorator for functions similar to the one we use to open files. But perhaps I am getting too ambitious and we should merge this first and see if we like it.  I'm fine with either approach.

**Comment by harryscholes** [comment]
Updates:
 - Put `random_seed` at the end of the function arguments
 - Created `check_random_state` function in `nx.utils.misc` that returns instances of `np.random.RandomState`
 - Add `@random_state()` decorator, which calls `check_random_state` on the value of `random_state` function arguments
 - Decorate `random_layout` and `fruchterman_reingold_layout`

**Comment by harryscholes** [comment]
Fixed lots of conflicts due to #2776 and #2681

**Comment by dschult** [comment]
Don't worry about the travis test failure in mac osx. They changed their build environment.

Minor things:   
- don't wrap the imports in try/except. (too much overhead and they don't cause errors if not in the module namespace.
- For sphinx, the docstring should start with a single line description (that means <80 chars too). Then a blank line and then (if reasonable) a paragraph defining what the terms mean in the single line description.   Blank lines separate sections. There are competing styles for docstrings, but take a look at [this one](http://www.sphinx-doc.org/en/stable/ext/example_numpy.html) from numpy and sphinx.

This is great stuff!  Thanks for chasing the bugs too!

**Comment by harryscholes** [comment]
Cheers Dan – I've incorporated your minor points and added tests for the `@random_state` decorator. All the tests are now passing.

**Comment by dschult** [comment]
OK... Is this all that is planned for this PR?  
If so, I'll take one more check through it and then merge.
:)

**Comment by harryscholes** [comment]
Your changes look good 😄

Yep I think this is all for this PR. In the future we can use these findings to address #1764 and related issues to replace `random.seed` with `random.Random` instances.
